examples/hitl/rearrange_v2/app_state_end_session.py

```diff
@@ -55,5 +63,58 @@ def sim_update(self, dt: float, post_sim_update_dict):
         self._elapsed_time += dt
 
     def _end_session(self):
-        # TODO: Data collection.
-        pass
+        session = self._session
+        if session is None:
+            print("Null session. Skipping S3 upload.")
+            return
+
+        # Finalize session.
+        if self._session.error == "":
+            session.success = True
+        session.session_recorder.end_session(self._session.error)
+
+        # Get data collection parameters.
+        try:
+            config = self._app_service.config
+            data_collection_config = config.rearrange_v2.data_collection
+            s3_path = data_collection_config.s3_path
+            if s3_path[-1] != "/":
+                s3_path += "/"
+            s3_subdir = "complete" if session.success else "incomplete"
+            s3_path += s3_subdir
+
+            # Use the port as a discriminator for when there are multiple concurrent servers.
+            output_folder_suffix = str(config.habitat_hitl.networking.port)
+            output_folder = f"output_{output_folder_suffix}"
+
+            output_file_name = data_collection_config.output_file_name
+            output_file = f"{output_file_name}.json.gz"
+
+        except Exception as e:
+            print(f"Invalid data collection config. Skipping S3 upload. {e}")
+            return
+
+        # Delete previous output directory
+        if os.path.exists(output_folder):
+            shutil.rmtree(output_folder)
```

Comment:
Just making sure this will not delete useful data?

Reply:
This only contains the `.json.gz` file. You could however change the path in the config to any directory 🤔